pkg/cmd/readyz/readyz.go

```
@@ -0,0 +1,187 @@
+}
+
+// Validate verifies the inputs.
+func (r *readyzOpts) Validate() error {
+	if len(r.targetEndpoint) == 0 {
+		return fmt.Errorf("missing required flag: --target")
+	}
+	if len(r.clientKeyFile) == 0 {
+		return errors.New("missing required flag: --key-file")
+	}
+	if len(r.clientCertFile) == 0 {
+		return errors.New("missing required flag: --cert-file")
+	}
+	if len(r.clientCACertFile) == 0 {
+		return errors.New("missing required flag: --cacert-file")
+	}
+
+	return nil
+}
+
+// Run contains the logic of the readyz command which checks the health of the etcd member
+func (r *readyzOpts) Run() error {
+	shutdownCtx, cancel := context.WithCancel(context.Background())
+	shutdownHandler := server.SetupSignalHandler()
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/readyz", r.getReadyzHandlerFunc(shutdownCtx))
+
+	addr := fmt.Sprintf("0.0.0.0:%d", r.listenPort)
+	klog.Infof("Listening on %s", addr)
+
+	server := &http.Server{
+		Addr:        addr,
+		Handler:     mux,
+		BaseContext: func(_ net.Listener) context.Context { return shutdownCtx },
+	}
+	go func() {
+		defer cancel()
+		<-shutdownHandler
+		klog.Infof("Received SIGTERM or SIGINT signal, shutting down server.")
+		server.Shutdown(shutdownCtx)
+	}()
+
+	c := net.ListenConfig{}
+	c.Control = permitAddressReuse
+	ln, err := c.Listen(shutdownCtx, "tcp", addr)
+	if err != nil {
+		return err
+	}
+	err = server.Serve(ln)
+	if err == http.ErrServerClosed {
+		err = nil
+	}
+	<-shutdownCtx.Done()
+	return err
+}
+
+func (r *readyzOpts) getReadyzHandlerFunc(ctx context.Context) http.HandlerFunc {
+	return func(w http.ResponseWriter, _ *http.Request) {
+		etcdClient, err := r.newETCD3Client(ctx, r.targetEndpoint)
+		if err != nil {
+			// TODO: The error message won't show up in the pod description if the probe fails
+			// e.g Warning  ProbeError  27s   kubelet  Readiness probe error: HTTP probe failed with statuscode: 503
+			// Find a way to get the error message to show there.
```

Seems like the kubelet's prober ignores the error message written to the body:
https://github.com/kubernetes/kubernetes/blob/56062f7f4fc56e038f9ab99042457183896c47bb/pkg/probe/http/http.go#L137-L138

It's available as a v4 log though, and we can alway query the readyz enpdpoint directly to confirm the error when it's failing so going to remove this TODO.

---

v(2) if you please.

---

> Seems like the kubelet's prober ignores the error message written to the body:
https://github.com/kubernetes/kubernetes/blob/56062f7f4fc56e038f9ab99042457183896c47bb/pkg/probe/http/http.go#L137-L138

our fork of kube includes body message snippets for certain namespaces.